Source/Core/Core/AchievementManager.cpp

```diff
@@ -129,15 +132,153 @@ void AchievementManager::LoadGameByFilenameAsync(const std::string& iso_path,
+  {
+    auto iter =
+        m_unlock_map.insert({m_game_data.achievements[ix].id, UnlockStatus{.game_data_index = ix}});
+    ActivateDeactivateAchievement(iter.first->first, enabled, unofficial, encore);
+  }
+}
+
+void AchievementManager::ActivateDeactivateLeaderboards()
+{
+  bool leaderboards_enabled = Config::Get(Config::RA_LEADERBOARDS_ENABLED);
+  for (u32 ix = 0; ix < m_game_data.num_leaderboards; ix++)
+  {
+    auto leaderboard = m_game_data.leaderboards[ix];
+    if (m_is_game_loaded && leaderboards_enabled && hardcore_mode_enabled)
+    {
+      rc_runtime_activate_lboard(&m_runtime, leaderboard.id, leaderboard.definition, nullptr, 0);
+    }
+    else
+    {
+      rc_runtime_deactivate_lboard(&m_runtime, m_game_data.leaderboards[ix].id);
+    }
+  }
+}
+
+void AchievementManager::ActivateDeactivateRichPresence()
+{
+  rc_runtime_activate_richpresence(
+      &m_runtime,
+      (m_is_game_loaded && Config::Get(Config::RA_RICH_PRESENCE_ENABLED)) ?
+          m_game_data.rich_presence_script :
+          "",
+      nullptr, 0);
+}
+
+void AchievementManager::DoFrame()
+{
+  if (m_is_game_loaded)
+  {
+    rc_runtime_do_frame(
+        &m_runtime,
+        [](const rc_runtime_event_t* runtime_event) {
+          AchievementManager::GetInstance()->AchievementEventHandler(runtime_event);
+        },
+        [](unsigned address, unsigned num_bytes, void* ud) {
+          return AchievementManager::GetInstance()->MemoryPeeker(address, num_bytes, ud);
+        },
+        nullptr, nullptr);
+    u64 current_time = Core::System::GetInstance().GetCoreTiming().GetTicks();
+    if (current_time - m_last_ping_time > SystemTimers::GetTicksPerSecond() * 120)
+      m_queue.EmplaceItem([this] { PingRichPresence(GenerateRichPresence()); });
+  }
+}
+
+unsigned AchievementManager::MemoryPeeker(unsigned address, unsigned num_bytes, void* ud)
+{
+  const rc_memory_regions_t* regions = rc_console_memory_regions(m_game_data.console_id);
+  for (unsigned int ix = 0; ix < regions->num_regions; ix++)
+  {
+    if (address >= regions->region[ix].start_address && address <= regions->region[ix].end_address)
+    {
+      address += (regions->region[ix].real_address - regions->region[ix].start_address);
+      break;
+    }
+  }
+  Core::CPUThreadGuard guard(Core::System::GetInstance());
+  switch (num_bytes)
+  {
+  case 1:
+    return PowerPC::MMU::HostTryReadU8(guard, address + 0x80000000)
+        .value_or(PowerPC::ReadResult<u8>(false, 0u))
+        .value;
+  case 2:
+    return PowerPC::MMU::HostTryReadU16(guard, address + 0x80000000)
+        .value_or(PowerPC::ReadResult<u16>(false, 0u))
+        .value;
+  case 4:
+    return PowerPC::MMU::HostTryReadU32(guard, address + 0x80000000)
+        .value_or(PowerPC::ReadResult<u32>(false, 0u))
+        .value;
+  case 8:
+    return PowerPC::MMU::HostTryReadU64(guard, address + 0x80000000)
+        .value_or(PowerPC::ReadResult<u64>(false, 0u))
+        .value;
+  default:
+    return 0u;
```

Comment:
May want to add an assert here or otherwise put a comment here if this is a valid path